neuralmonkey/encoders/recurrent.py

```diff
@@ -0,0 +1,215 @@
+            fw_cell, bw_cell, embedded,
+            sequence_length=self.input_sequence.lengths,
+            dtype=tf.float32)
+
+    @tensor
+    def states(self) -> tf.Tensor:
+        # pylint: disable=unsubscriptable-object
+        return tf.concat(self.bidirectional_rnn[0], 2)
+        # pylint: enable=unsubscriptable-object
+
+    @tensor
+    def encoded(self) -> tf.Tensor:
+        # pylint: disable=unsubscriptable-object
+        if self.rnn_cell_str == "GRU":
+            return tf.concat(self.bidirectional_rnn[1], 1)
+        elif self.rnn_cell_str == "LSTM":
+            # TODO is "h" what we want?
+            final_states = [state.h for state in self.bidirectional_rnn[1]]
+            return tf.concat(final_states, 1)
+        # pylint: enable=unsubscriptable-object
+
+    @tensor
+    def _attention_tensor(self) -> tf.Tensor:
+        return dropout(self.states, self.dropout_keep_prob, self.train_mode)
+
+    @tensor
+    def _attention_mask(self) -> tf.Tensor:
+        # TODO tohle je proti OOP prirode
+        return self.input_sequence.mask
+
+    @tensor
+    def states_mask(self) -> tf.Tensor:
+        return self.input_sequence.mask
+
+    def _rnn_cells(self) -> RNNCellTuple:
+        """Return the graph template to for creating RNN memory cells"""
+        return(RNN_CELL_TYPES[self.rnn_cell_str](self.rnn_size),
+               RNN_CELL_TYPES[self.rnn_cell_str](self.rnn_size))
+
+    def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
+        fd = self.input_sequence.feed_dict(dataset, train)
+        fd[self.train_mode] = train
+
+        return fd
+
+
+class SentenceEncoder(RecurrentEncoder):
+    # pylint: disable=too-many-arguments,too-many-locals
+    def __init__(self,
+                 name: str,
+                 vocabulary: Vocabulary,
+                 data_id: str,
+                 embedding_size: int,
+                 rnn_size: int,
+                 max_input_len: int = None,
+                 dropout_keep_prob: float = 1.0,
+                 rnn_cell: str = "GRU",
+                 attention_type: type = None,
+                 attention_fertility: int = 3,
+                 attention_state_size: int = None,
+                 save_checkpoint: str = None,
+                 load_checkpoint: str = None) -> None:
+        """Create a new instance of the sentence encoder. """
+
+        # TODO Think this through.
+        s_ckp = "input_{}".format(save_checkpoint) if save_checkpoint else None
+        l_ckp = "input_{}".format(load_checkpoint) if load_checkpoint else None
+
+        # TODO! Representation runner needs this. It is not simple to do it in
+        # recurrent encoder since there may be more source data series. The
+        # best way could be to enter the data_id parameter manually to the
+        # representation runner
+        self.data_id = data_id
+
+        input_sequence = EmbeddedSequence(
+            name="{}_input".format(name),
```

Comment:
Ja mel za to, ze zpetna kompatibilita se uzavrela poslednim releasem a ted je moment kdy ji muzem opustit a trochu ten kod zlepsit.